### PR TITLE
Visualization tools for ApplyForceTorque

### PR DESCRIPTION
**Pull request by Henrique-BO**
# 🎉 New feature

Requires #2014 

## Summary

Adds visualization tools to the new ApplyForceTorque GUI plugin. This allows the user to see the force (represented by an arrow) and the torque (represented by a circle) displayed on the scene at their application point before applying them.

When one of the vectors (force or torque) is clicked, a rotation tool (represented by a gizmo) is displayed. This allows the user to freely rotate the selected vector. Still WIP.

TODO:
- [x] Improve torque visualization
- [x] Implement rotation tool for the vectors (like in [Classic](http://classic.gazebosim.org/tutorials?tut=apply_force_torque#Rotationtool))

## Test it

Open any test world.

Load the Apply Force Torque plugin from the plugin dropdown.

Select a model or link and specify a force or torque to see its direction on the scene.

Click on the force or torque vector to display the rotation tool, and click and drag the circles to rotate.

image

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by arjo129** [comment]
@Henrique-BO this is really cool. It would be nice if we could extract the rendering logic out to some common area so it can be shared by other plugins. #1898 seeks to do visualizations but on arbitrary objects. It would be nice to have a consistent vsiaul language between these two PRs. Perhaps if you were to create a new utility class called "Force Visuallization" and make the material configurable, we would be able to reuse this code in #1898.

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc [hunk 1]
```diff
@@ -165,25 +228,59 @@ void ApplyForceTorque::LoadConfig(const tinyxml2::XMLElement */*_pluginElem*/)
 /////////////////////////////////////////////////
 bool ApplyForceTorque::eventFilter(QObject *_obj, QEvent *_event)
 {
-  std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
+  // std::lock_guard<std::mutex> lock(this->dataPtr->mutex);
```
Comment (ahcorde) [inline src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:231]: remove this line

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc [hunk 2]
```diff
@@ -519,14 +673,204 @@ void ApplyForceTorquePrivate::OnRender()
...
+
+    /// get start and end pos in world frame from 2d point
+    math::Vector3d pos = this->linkWorldPose.Pos() + this->inertialPos +
+      this->linkWorldPose.Rot().RotateVector(this->offset);
+    double d = pos.Dot(axis);
+    math::Planed plane(axis, d);
+
+    math::Vector3d startPos;
+    this->ray->SetFromCamera(this->camera, start);
+    if (auto v = plane.Intersection(
```
Comment (ahcorde) [inline src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:843]: nit: use `{}`

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc [hunk 3]
```diff
@@ -519,14 +673,204 @@ void ApplyForceTorquePrivate::OnRender()
...
+      this->ray->Origin(), this->ray->Direction()))
+      startPos = *v;
+    else
+      return;
+
+    math::Vector3d endPos;
+    this->ray->SetFromCamera(this->camera, end);
+    if (auto v = plane.Intersection(
```
Comment (ahcorde) [inline src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:851]: nit: use `{}`

**Comment by arjo129** [comment]
Sounds awesome!

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc [hunk 4]
```diff
@@ -398,5 +566,343 @@ void ApplyForceTorquePrivate::PublishWrench(bool _applyForce, bool _applyTorque)
...
+  {
+    // Events with false should only be sent once
+    if (!this->blockOrbit)
+      this->sendBlockOrbit = false;
+
+    gz::gui::events::BlockOrbit blockOrbitEvent(this->blockOrbit);
+    gz::gui::App()->sendEvent(
+        gz::gui::App()->findChild<gz::gui::MainWindow *>(),
+        &blockOrbitEvent);
+  }
```
Comment (azeey) [inline src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:624]: Why does this plugin need to send `BlockOribit`? Does the user interact with it in the 3D scene?

Edit: Ah, after reading the description, I see that there's supposed to be some interaction. However, this is not working for me. Anytime I interact with the 3D scene, the Orbit view is taking over.

**Review by azeey — COMMENTED** [review]
Looks great! I wasn't able to get the 3D Scene interaction to work. Can you try it again? Otherwise, just a few minor comments.

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc [hunk 5]
```diff
@@ -398,5 +566,343 @@ void ApplyForceTorquePrivate::PublishWrench(bool _applyForce, bool _applyTorque)
...
+    yVis->SetWorldRotation(gizmoRot *
+      math::Quaterniond(yRot) * math::Quaterniond(yRotOffset));
+
+    rendering::VisualPtr zVis = this->gizmoVisual->ChildByAxis(
+      rendering::TransformAxis::TA_ROTATION_Z);
+    math::Vector3d zRot(0, 0, atan2(eye.Y(), eye.X()));
+    zVis->SetWorldRotation(gizmoRot * math::Quaterniond(zRot));
+
+    rendering::VisualPtr circleVis = this->gizmoVisual->ChildByAxis(
+      rendering::TransformAxis::TA_ROTATION_Z << 1);
```
Comment (azeey) [inline src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:719]: FYI: https://github.com/gazebosim/gz-rendering/issues/880

**Comment by Henrique-BO** [comment]
Changes addressed in b5cafc3.

The 3D scene interaction is working on my end, like this:

https://github.com/gazebosim/gz-sim/assets/49255495/425f466a-e4dd-4de4-aaed-b23a456e58c3

Is this not working for you?

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc (hunk 5 above)
Comment (iche033) [reply src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:723]: This looks like same logic in `TransformController` class in gz-rendering. I'll provide a function in `GizmoVisual` class in gz-rendering to take care of this logic so we can avoid duplicating the code here. The changes are ABI breaking. So I would suggest leaving the changes like this in gz-sim7 (maybe with a todo note) and we can remove the code in `main`

Comment (azeey) [reply src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:723]: Sounds good!

Comment (iche033) [reply src/gui/plugins/apply_force_torque/ApplyForceTorque.cc:723]: https://github.com/gazebosim/gz-rendering/pull/882

File: src/gui/plugins/apply_force_torque/ApplyForceTorque.cc [hunk 6]
```diff
@@ -398,5 +567,346 @@ void ApplyForceTorquePrivate::PublishWrench(bool _applyForce, bool _applyTorque)
...
+    this->scene = rendering::sceneFromFirstRenderEngine();
+    if (!this->scene)
+    {
+      return;
+    }
+
+    for (unsigned int i = 0; i < this->scene->NodeCount(); ++i)
+    {
+      auto cam = std::dynamic_pointer_cast<rendering::Camera>(
+        this->scene->NodeByIndex(i));
```
[the remark this answers is not in the document]
Comment (Henrique-BO) [reply src/gui/plugins/apply_force_torque/ApplyForceTorque.cc]: Good point, changed it in d6afb10

**Comment by Henrique-BO** [comment]
Good point, I added a message to the GUI in 419b34f

**Review by iche033 — APPROVED** [review]
nice, the tool works well for me

**Review by azeey — APPROVED** [review]

**Pull request merged**